src/smt/tactic/smt_tactic_core.cpp

```diff
@@ -101,10 +103,26 @@ class smt_tactic : public tactic {
 
 
     void collect_statistics(statistics & st) const override {
-        if (m_ctx)
-            m_ctx->collect_statistics(st); // ctx is still running...
-        else
-            st.copy(m_stats);
+        // Use compare_exchange to safely extract the context pointer
+        // If we win the race, we're responsible for deallocation
+        smt::kernel* ctx = m_ctx.load();
+        if (ctx) {
+            // Try to atomically swap the context with nullptr
+            // If successful, we own the context and must deallocate it
+            if (m_ctx.compare_exchange_strong(ctx, nullptr)) {
+                try {
+                    ctx->collect_statistics(st);
+                } catch (...) {
+                    // If exception occurs, fall back to cached stats
+                    st.copy(m_stats);
+                }
+                // We won the race, so we're responsible for cleanup
+                dealloc(ctx);
```

Comment:
this is wrong. The context should only be deallocated after search has completed.

Reply:
maybe just restore the behavior of collect_statistics that worked on m_ctx directly and catch an exception if thrown?